src/drivers/legos/hid_report.rs

```diff
@@ -0,0 +1,190 @@
+    #[packed_field(bits = "10")]
+    pub d_trigger_l: bool,
+    #[packed_field(bits = "9")]
+    pub rb: bool,
+    #[packed_field(bits = "8")]
+    pub d_trigger_r: bool,
+    #[packed_field(bits = "7")]
+    pub legion: bool,
+    #[packed_field(bits = "6")]
+    pub quick_access: bool,
+    #[packed_field(bits = "5")]
+    pub thumb_l: bool,
+    #[packed_field(bits = "4")]
+    pub thumb_r: bool,
+    #[packed_field(bits = "3")]
+    pub up: bool,
+    #[packed_field(bits = "2")]
+    pub down: bool,
+    #[packed_field(bits = "1")]
+    pub left: bool,
+    #[packed_field(bits = "0")]
+    pub right: bool,
+
+    #[packed_field(bytes = "4")]
+    pub l_stick_x: i8,
+    #[packed_field(bytes = "5")]
+    pub l_stick_y: i8,
+    #[packed_field(bytes = "6")]
+    pub r_stick_x: i8,
+    #[packed_field(bytes = "7")]
+    pub r_stick_y: i8,
+
+    // byte 8
+    #[packed_field(bits = "71")]
+    pub rpad_touching: bool,
+    // byte 9
+    #[packed_field(bits = "79")]
+    pub rpad_tap: bool,
+
+    #[packed_field(bytes = "10")]
+    pub touch_x: i8,
+    #[packed_field(bytes = "11")]
+    pub touch_y: i8,
+
+    #[packed_field(bytes = "12")]
+    pub a_trigger_l: u8,
+    #[packed_field(bytes = "13")]
+    pub a_trigger_r: u8,
+}
+
+impl XInputDataReport {
+    /// Determines if the current report matches the bad data report.
+    pub fn is_bad_data(&self) -> bool {
+        if self.matches(&bad_data) {
+            return true;
+        }
+        false
+    }
+
+    /// Determines if the current report matches the given report.
+    fn matches(&self, data: &XInputDataReport) -> bool {
+        if self.y1 == data.y1
+            && self.y2 == data.y2
+            && self.view == data.view
+            && self.menu == data.menu
+            && self.a == data.a
+            && self.b == data.b
+            && self.x == data.x
+            && self.y == data.y
+            && self.lb == data.lb
+            && self.d_trigger_l == data.d_trigger_l
+            && self.rb == data.rb
+            && self.d_trigger_r == data.d_trigger_r
+            && self.legion == data.legion
+            && self.quick_access == data.quick_access
+            && self.thumb_l == data.thumb_l
+            && self.thumb_r == data.thumb_r
+            && self.up == data.up
+            && self.down == data.down
+            && self.left == data.left
+            && self.right == data.right
+            && self.l_stick_x == data.l_stick_x
+            && self.l_stick_y == data.l_stick_y
+            && self.r_stick_x == data.r_stick_x
+            && self.r_stick_y == data.r_stick_y
+            && self.rpad_touching == data.rpad_touching
+            && self.rpad_tap == data.rpad_tap
+            && self.touch_x == data.touch_x
+            && self.touch_y == data.touch_y
+            && self.a_trigger_l == data.a_trigger_l
+            && self.a_trigger_r == data.a_trigger_r
+        {
+            true
+        } else {
+            false
+        }
+    }
```

Comment:
Because this struct implements `PartialEq`, we should be able to compare data directly:

```suggestion
    /// Determines if the current report matches the bad data report.
    pub fn is_bad_data(&self) -> bool {
        *self == bad_data
    }
```